### PR TITLE
WIP: Add kubernetes-cluster receiver

### PR DESCRIPTION
**Pull request by asuresh4**
A receiver to watch for changes to resources in a Kubernetes cluster and collect metrics and metadata.

### CODE REVIEW

File: receiver/kubernetesclusterreceiver/watcher.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// getMetricsFromCache returns metricsCache stored in the cache at a given point in time
+func (rw *resourceWatcher) getMetricsFromCache() []*consumerdata.MetricsData {
+	var out []*consumerdata.MetricsData
+
+	rw.metricsCollector.Lock()
+	defer rw.metricsCollector.Unlock()
+
+	for _, metricdata := range rw.metricsCollector.metricsCache {
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/watcher.go]: Do we need to construct this list or can we just call ConsumeMetric data for all values in the map? Maybe have a processMetricsCache that takes a function and in that function parameter can do the call to ConsumeMetrics.

Though I guess that will result in the metricsCache being locked longer if ConsumeMetrics is slow 🤔. I don't know the performance characteristics of the pipeline. Maybe @tigrannajaryan or @bogdandrutu can shed some light. 

As is though it can be improved by preallocating the array upfront:
```go
out = make([]*consumerdata.MetricsData, len(rw.metricsCollector.metricsCache))
```

Also accessing `rw.metricsCollector.metricsCache` feels like it's reaching into some other objects internals, might be a signal that it needs some refactoring there.

File: receiver/kubernetesclusterreceiver/config.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+)
+
+// Config defines configuration for kubernetes cluster receiver.
+type Config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
+
+	// Collection interval for metrics
+	CollectionInterval time.Duration `mapstructure:"collection_interval"`
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply receiver/kubernetesclusterreceiver/config.go]: Yeah, maybe interval is too ambiguous. collection_interval is fine with me.

File: receiver/kubernetesclusterreceiver/nodes.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	corev1 "k8s.io/api/core/v1"
+)
+
+func getMetricsForNode(node *corev1.Node, nodeConditionTypesToReport []string) *resourceMetrics {
+	metrics := make([]*metricspb.Metric, 0)
+	for _, nodeConditionTypeValue := range nodeConditionTypesToReport {
+		nodeConditionMetric := fmt.Sprintf("kubernetes/node/condition_%s", strcase.ToSnake(nodeConditionTypeValue))
+		v1NodeConditionTypeValue := corev1.NodeConditionType(nodeConditionTypeValue)
+
+		metrics = []*metricspb.Metric{
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/nodes.go:21]: Is this supposed to be appending to metrics?

File: receiver/kubernetesclusterreceiver/watcher.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	rw.addOrUpdateResource(obj)
+}
+
+func (rw *resourceWatcher) onDelete(obj interface{}) {
+
+	key, err := getUIDForObject(obj.(runtime.Object))
+
+	if err != nil {
+		rw.logger.Error(
+			fmt.Sprintf("failed to remove from metric cache: %s", err),
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/watcher.go:125]: there's a zap.Error field

File: receiver/kubernetesclusterreceiver/watcher.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+	if rm == nil || rm.resource == nil || rm.metrics == nil {
+		return
+	}
+
+	key, err := getUIDForObject(obj.(runtime.Object))
+
+	if err != nil {
+		rw.logger.Error(
+			fmt.Sprintf("failed to update metric cache: %s", err),
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/watcher.go:191]: Same

File: receiver/kubernetesclusterreceiver/collector.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/utils"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type metricsCollector struct {
+	sync.RWMutex
+	metricsCache map[types.UID]*consumerdata.MetricsData
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/collector.go:19]: MetricsData doesn't need to be a pointer since it's a small struct with pointers within it. (ConsumeMetrics doesn't take it as a pointer either). If it's a small struct with just a couple pointers in it probably doesn't need to be referenced as a pointer. I haven't figured out what the definite cutoffs for pointer vs. not but can look to existing code to see how they're using it.

Decent rule of thumb I've read is that if the struct is the size of a cache line or smaller then it might be a good candidate to not reference as a pointer.

File: receiver/kubernetesclusterreceiver/collector.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+type resourceMetrics struct {
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/collector.go:22]: If this is unlikely to get bigger this is probably a good example of something small enough to not use as pointer.

File: receiver/kubernetesclusterreceiver/cronjobs.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+var activeJobs = &metricspb.MetricDescriptor{
+	Name:        "kubernetes/cronjob/active_jobs",
+	Description: "The number of actively running jobs for a cronjob",
+	Unit:        "1",
+	Type:        metricspb.MetricDescriptor_GAUGE_INT64,
+}
+
+func getMetricsForCronJob(cj *batchv1beta1.CronJob) *resourceMetrics {
+	metrics := make([]*metricspb.Metric, 0)
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/cronjobs.go:19]: Don't need this, can just put it inline with the return statement.

File: receiver/kubernetesclusterreceiver/pods.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+var podPhaseMetric = &metricspb.MetricDescriptor{
+	Name:        "kubernetes/pod/phase",
+	Description: "Current phase of the pod (1 - Pending, 2 - Running, 3 - Succeeded, 4 - Failed, 5 - Unknown)",
+	Unit:        "",
+	Type:        metricspb.MetricDescriptor_GAUGE_INT64,
+}
+
+func getResourceMetricsForPod(pod *corev1.Pod) *resourceMetrics {
+	metrics := make([]*metricspb.Metric, 0)
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/pods.go:27]: Same here

File: receiver/kubernetesclusterreceiver/replica.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"fmt"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/utils"
+)
+
+func getReplicaMetrics(resource string, desired, available int32) []*metricspb.Metric {
+	out := []*metricspb.Metric{
```
Comment (jrcamp) [inline receiver/kubernetesclusterreceiver/replica.go:11]: Would just return

File: receiver/kubernetesclusterreceiver/nodes.go (hunk 3 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/nodes.go]: Yes, thanks for catching it. I've updated it.

File: receiver/kubernetesclusterreceiver/watcher.go (hunk 4 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/watcher.go]: updated

File: receiver/kubernetesclusterreceiver/watcher.go (hunk 5 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/watcher.go]: updated

File: receiver/kubernetesclusterreceiver/collector.go (hunk 6 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/collector.go]: Thanks for pointing this out. I normally tend to reference structs as pointers when in doubt. But I'll keep this in mind going forward. Also, updated this.

File: receiver/kubernetesclusterreceiver/cronjobs.go (hunk 8 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/cronjobs.go]: updated

File: receiver/kubernetesclusterreceiver/pods.go (hunk 9 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/pods.go]: updated

File: receiver/kubernetesclusterreceiver/replica.go (hunk 10 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/replica.go]: updated

File: receiver/kubernetesclusterreceiver/collector.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	tags          map[string]bool
+}
+
+// getGenericMetadata is responsible for collecting metadata from K8s resources that
+// live on v1.ObjectMeta.
+func getGenericMetadata(om *v1.ObjectMeta, resourceType string) *resourceMetadata {
+	properties, tags := utils.GetPropertiesAndTagsFromLabels(om.Labels)
+
+	properties["kubernetes_workload_name"] = om.Name
+	properties["kubernetes_workload"] = resourceType
```
[the remark this answers is not in the document]
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/collector.go]: Moved most of the keys as constants. I targeted the ones that are being used in more than one place and the keys such as ID or name of resources.

File: receiver/kubernetesclusterreceiver/containers.go [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+	Name:        "kubernetes/container/cpu_limit",
+	Description: "Maximum CPU limit set for the container",
+	Unit:        "",
+	Type:        metricspb.MetricDescriptor_GAUGE_INT64,
+}
+
+var containerMemoryRequestMetric = &metricspb.MetricDescriptor{
+	Name:        "kubernetes/container/memory_request",
+	Description: "Memory requested for the container",
+	Unit:        "",
```
[the remark this answers is not in the document]
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/containers.go]: Yes, updated most of the current metrics to use correct units. However, I'm not sure how we should handle metrics that indicate, for example, the phase of a Pod. In that case the metricization is driven by an enumeration where different values indicate the phase of the Pod. (1 - Pending, 2 - Running, 3 - Succeeded, 4 - Failed, 5 - Unknown). Do we have a convention to handle metrics like these?

File: .circleci/config.yml [hunk 13]
```diff
@@ -1,4 +1,32 @@
-version: 2
+version: 2.1
```
Comment (tigrannajaryan) [inline .circleci/config.yml]: I assume this changes are not relevant for the PR, so I am skipping build-related changes.

File: cmd/otelcontribcol/components.go [hunk 14]
```diff
@@ -72,26 +75,27 @@ func components() (config.Factories, error) {
...
 		&carbonexporter.Factory{},
 		&honeycombexporter.Factory{},
+		&jaegerthrifthttpexporter.Factory{},
 		&lightstepexporter.Factory{},
 	}
 	for _, exp := range factories.Exporters {
 		exporters = append(exporters, exp)
 	}
-	factories.Exporters, err = exporter.Build(exporters...)
+	factories.Exporters, err = component.MakeExporterFactoryMap(exporters...)
```
Comment (tigrannajaryan) [inline cmd/otelcontribcol/components.go]: I see changes that seem unrelated to adding the receiver. Can you fix the PR to exclude them to make reviewing relevant parts easier?

File: receiver/kubernetesclusterreceiver/collector.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	k8SKeyReplicaSetUID            = "k8s.replicaset.uid"
+	k8SKeyReplicationControllerUID = "k8s.replicationcontroller.uid"
+
+	// Resource labels keys for Name
+	k8SKeyCronJobName               = "k8s.cronjob.name"
+	k8SKeyNodeName                  = "k8s.node.name"
+	k8SKeyDeploymentName            = "k8s.deployment.name"
+	k8SKeyJobName                   = "k8s.job.name"
+	k8SKeyReplicaSetName            = "k8s.replicaset.name"
+	k8SKeyReplicationControllerName = "k8s.replicationcontroller.name"
```
Comment (tigrannajaryan) [inline receiver/kubernetesclusterreceiver/collector.go:47]: Nice, very useful to have this in one place.

OpenTelemetry defines conventions for K8S attributes: https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions#kubernetes

Looks like the only intersecting attribute in this list "k8s.deployment.name" and it luckily matches what OpenTelemetry does. I wonder if we should submit a proposal to OpenTelemetry to extend the convensions to include some/all of these?

Do note that we have conventions defined in "github.com/open-telemetry/opentelemetry-collector/translator/conventions/opentelemetry.go" so if we decide to propose this to OpenTelemetry then we will to add them to that file and import here.

File: receiver/kubernetesclusterreceiver/collector.go [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+	// Resource labels for container
+	containerKeyID       = "container.id"
+	containerKeySpecName = "container.spec.name"
+
+	// Keys for K8s properties
+	k8SKeyWorkLoad     = "kubernetes_workload"
```
Comment (tigrannajaryan) [inline receiver/kubernetesclusterreceiver/collector.go:54]: Is there a reason this follows a different naming convention and uses "kubernetes_" as a prefix instead of "k8s."?

File: receiver/kubernetesclusterreceiver/factory.go [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/config/configerror"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"go.uber.org/zap"
+)
+
+const (
+	// Value of "type" key in configuration
+	typeStr = "kubernetes-cluster"
```
Comment (tigrannajaryan) [inline receiver/kubernetesclusterreceiver/factory.go:29]: For consistency please use underscore instead of dashes in config keys.

Also, I think we should go with "k8s_cluster". It is shorter and more consistent with "k8s_tagger" processor that already exists.

File: receiver/kubernetesclusterreceiver/pods.go [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+	return tags
+}
+
+// getPodCronJobProperties returns properties of a CronJob associated with the pod
+func getPodCronJobProperties(cronJobRef *v1.OwnerReference) map[string]string {
+	return map[string]string{
+		k8SKeyWorkLoad:     "CronJob",
+		k8SKeyWorkLoadName: cronJobRef.Name,
+		"cronJob":          cronJobRef.Name,
+		"cronJob_uid":      string(cronJobRef.UID),
```
Comment (tigrannajaryan) [inline receiver/kubernetesclusterreceiver/pods.go:228]: Is this mixed case usage in "cronJob_uid" intentional? Perhaps stick to either camelCase or underscore_case?

Also, can we define these as consts as well?

File: receiver/kubernetesclusterreceiver/factory.go (hunk 17 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/factory.go:29]: I agree, will update.

File: .circleci/config.yml (hunk 13 above)
Comment (asuresh4) [reply .circleci/config.yml]: This likely was pulled in when had rebased this PR and our fork with upstream. This is not related to the this PR, I'll fix the PR to remove these changes.

File: receiver/kubernetesclusterreceiver/factory.go (hunk 17 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/factory.go:29]: Will update the type to be `k8s_cluster`

File: receiver/kubernetesclusterreceiver/collector.go (hunk 16 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/collector.go:54]: For resource label, I followed https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions#kubernetes.

However, for properties, I stuck to conventions in the agent monitor. But now that I think about it more, I guess these could also be prefixed the same.

File: receiver/kubernetesclusterreceiver/collector.go (hunk 15 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/collector.go]: Yes, I think it would be useful to extend the list "github.com/open-telemetry/opentelemetry-collector/translator/conventions/opentelemetry.go". Especially because these keys pertain to different resource types in k8s and follow the conventions. Would you like me to include these constants to the core? @tigrannajaryan @bogdandrutu

File: receiver/kubernetesclusterreceiver/collector.go (hunk 7 above)
Comment (asuresh4) [reply receiver/kubernetesclusterreceiver/collector.go]: This struct hopefully be temporary. Once we have the new OTLP representation (https://github.com/open-telemetry/opentelemetry-proto/blob/master/gen/go/metrics/v1/metrics.pb.go#L104), these will be replaced.

**Comment by asuresh4** [comment]
Closing in favor of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/175
